### PR TITLE
[SYCL] [NATIVECPU] Implement generic atomic store for generic target

### PR DESCRIPTION
**Pull request by PietroGhg**
Implements `__spirv_AtomicStore` similarly to https://github.com/intel/llvm/pull/13249. Note that the `IMPL` macro has been extended to take in a `SUB` parameter, similarly to what happens for [amdgcn](https://github.com/intel/llvm/blob/a5a0e1296269195de90949537597b2788bb5e836/libclc/amdgcn-amdhsa/libspirv/atomic/atomic_store.cl#L13) and [ptx](https://github.com/intel/llvm/blob/a5a0e1296269195de90949537597b2788bb5e836/libclc/ptx-nvidiacl/libspirv/atomic/atomic_store.cl#L39).

### CODE REVIEW

File: libclc/generic/libspirv/atomic/atomic_store.cl [hunk 1]
```diff
@@ -29,26 +29,27 @@ _Z19__spirv_AtomicStorePU3AS3fN5__spv5Scope4FlagENS1_19MemorySemanticsMask4FlagE
...
+    }                                                                                                                             \
   }
 
-#define IMPL_AS(TYPE, TYPE_MANGLED, PREFIX, BYTE_SIZE) \
-IMPL(TYPE, TYPE_MANGLED, global, AS1, PREFIX, BYTE_SIZE) \
-IMPL(TYPE, TYPE_MANGLED, local, AS3, PREFIX, BYTE_SIZE)
+#define IMPL_AS(TYPE, TYPE_MANGLED, PREFIX, BYTE_SIZE)                         \
+  IMPL(TYPE, TYPE_MANGLED, global, U3AS1, 1, PREFIX, BYTE_SIZE)                \
+  IMPL(TYPE, TYPE_MANGLED, local, U3AS3, 1, PREFIX, BYTE_SIZE)                 \
+  IMPL(TYPE, TYPE_MANGLED, , , 0, PREFIX, BYTE_SIZE)
```
Comment (frasercrmck) [inline libclc/generic/libspirv/atomic/atomic_store.cl:52]: Do we need to be careful about a generic target not supporting the generic address space? If the target doesn't support this address space, then would an undecorated pointer not technically be in the `private` address space? Then the mangling of the function name might not necessarily match the mangling of its pointer arguments.

Sorry if this also applies to the atomic load patch too - I may not have been thinking clearly :wink:

Comment (PietroGhg) [reply libclc/generic/libspirv/atomic/atomic_store.cl:52]: Yeah I think you are right, currently I think that `Atomic*` operations in `generic` are used only by Native CPU, which also uses the `x86_64-unknown-linux` target folder. Do you think we should move all the `atomic/atomic_*.cl` files there and leave only the `loadstore_helpers` LLVM files in `generic` ?

Comment (frasercrmck) [reply libclc/generic/libspirv/atomic/atomic_store.cl:52]: Good question. My instinct is to keep generic things in the generic folder, even if in practice only one target uses them. I suppose we could argue about whether hard-coding address spaces is truly generic, but I'm sure we can fix that somehow.

So unless we've got no way of expressing these builtins in a generic way, I vote to keep them here. We can use the `_CLC_GENERIC_AS_SUPPORTED` (not merged yet - see https://github.com/intel/llvm/pull/13423) define to keep it as generic as possible.

Comment (PietroGhg) [reply libclc/generic/libspirv/atomic/atomic_store.cl:52]: Ok cool, I'll wait for https://github.com/intel/llvm/pull/13423 to get merged and use that macro for this PR (and the load too)

Comment (PietroGhg) [reply libclc/generic/libspirv/atomic/atomic_store.cl]: Hi @frasercrmck, I've changed the PR to use `_CLC_GENERIC_AS_SUPPORTED`, can you take another look? Thanks

File: libclc/generic/libspirv/atomic/atomic_load.cl [hunk 2]
```diff
@@ -43,6 +42,21 @@ IMPL_AS(long, l, , 8)
 IMPL_AS(unsigned long, m, u, 8)
 #endif
 
+#ifdef _CLC_GENERIC_AS_SUPPORTED
```
Comment (frasercrmck) [inline libclc/generic/libspirv/atomic/atomic_load.cl:45]: This should be `#if` - it's always defined (assuming the header is included) but is either 1 or 0

File: libclc/generic/libspirv/atomic/atomic_store.cl [hunk 3]
```diff
@@ -58,6 +58,21 @@ IMPL_AS(long, l, , 8)
 IMPL_AS(unsigned long, m, u, 8)
 #endif
 
+#ifdef _CLC_GENERIC_AS_SUPPORTED
```
Comment (frasercrmck) [inline libclc/generic/libspirv/atomic/atomic_store.cl:61]: same here - `#if`

**Review by frasercrmck — APPROVED** [review]
Looks good, thanks

**Comment by PietroGhg** [comment]
@frasercrmck I've just realized that after https://github.com/intel/llvm/pull/13109, `_CLC_GENERIC_AS_SUPPORTED` will actually be `0` for Native CPU, which means that we the current use of that macro the overloads will not be available on Native CPU, which kinda defies the purpose of this PR, so I am not really sure what to do here

**Review by frasercrmck — APPROVED** [review]
LGTM

**Comment by PietroGhg** [comment]
@intel/llvm-gatekeepers this looks ready to merge, thank you

**Pull request merged**